blog/reports/reports.md

```diff
@@ -0,0 +1,473 @@
+
+## System Architecture
+
+The reporting system isn't just bolted on as an afterthought - it's built into Magistrala's core architecture. Here's what you're working with:
+
+### Core Service Components
+
+1. **Report Service** : This is the heart of the operation. It handles all the business logic for creating, updating, and generating reports. The service manages everything from parsing time expressions like `now()-1h` to orchestrating the entire report generation pipeline. It communicates with the readers service via gRPC to fetch historical IoT data, applies aggregations (MIN, MAX, AVG, SUM, COUNT), and coordinates with the template engine for output formatting. The gRPC client connection allows for efficient, type-safe communication when querying large datasets.
+
+2. **Repository Layer** : All your report configurations, schedules, and metadata live here in PostgreSQL. This layer handles CRUD operations for report configs, stores custom templates, and manages the scheduling state. It's designed to handle the multi-tenant nature of Magistrala, so each domain's reports stay completely isolated.
+
+3. **API Layer** : The REST endpoints that actually don't make you want to pull your hair out. This layer handles authentication, request validation, and all the HTTP routing. It includes proper error handling and follows Magistrala's consistent API patterns, so if you've used other parts of the platform, this will feel familiar.
+
+4. **Scheduler Component**: Built on top of a ticker system, this handles the "set it and forget it" functionality. It wakes up periodically, checks for reports that need to run, and kicks off the generation process. The scheduler is smart enough to handle failures and won't spam you if something goes wrong.
+
+5. **Template Engine & PDF Generation**: This is what makes your PDFs look professional instead of like something from 1995. The system uses a two-part approach: first, it processes your data through Go HTML templates with custom styling, then sends the rendered HTML to **Gotenberg** - a dedicated microservice that converts HTML to clean PDF output. Gotenberg runs as a separate Docker container and provides a stateless API for document conversion. The engine validates templates to prevent security issues and the whole pipeline is designed for high-throughput PDF generation.
+
+6. **Email Integration**: Uses Magistrala's built-in emailer service to handle SMTP delivery. It validates recipient addresses, handles attachments properly, and includes retry logic for when email servers are being temperamental. The integration respects domain policies so you can't accidentally send internal data to external addresses.
+
+### The Data Journey
+
+```
+Your IoT Devices → Message Brokers → Readers Service → Reports Service → Reports
+```
+
+What's nice is that the reporting service doesn't reinvent the wheel. It hooks into the existing readers infrastructure, so you're querying the same data store your other services use.
```

Comment:
I am assuming a reader may not know what the readers service in magistrala is, you can give briefcontext into that